diff/walking/differ.go

```diff
@@ -70,12 +70,16 @@ func (s *walkingDiff) Compare(ctx context.Context, lower, upper []mount.Mount, o
 	}
 
 	var isCompressed bool
-	switch config.MediaType {
```

Comment:
do we need to validate the mediatype if it is not Gzip? If I understand correctly, this patch is only used for Gzip, right? If so, I think we should limit the usage here.

Reply:
I think we don't need that because keeping the consistency between `config.MediaType` and `config.Compressor` is the caller's responsibility and we can leverage this option for other compressions (e.g. zstd) as well.

cc @tonistiigi

Reply:
I just worry about the invalid data can be spread out by bug or mistake. For the patch, I think the extension should be in  small scope.

Reply:
Thank you for the quick review. Fixed the patch to keep the scope narrower.

Reply:
I'm not quite sure what this behavior used to be but the logic that would make sense to me would be that if `Compressor` is set, then the error case `errors.Wrapf(errdefs.ErrNotImplemented, "unsupported diff media type: %v", config.MediaType)` does not appear if mediatype is unknown and any mediatype is allowed. Currently, if we start to use this for zstd with the custom compressor we would still need to pass mediatype as gzip that is likely ignored by the `content.Store` implementation but may not be for custom implementation(eg. for a pusher). Plus caller would need to modify the returned descriptor back to `zstd`.

Reply:
I agree with that the differ should support zstd as well. So I think we can have a separated PR for explicitly expand the scope of the differ for zstd.

Reply:
@tonistiigi @ktock I agree with supporting zstd. I think the option should be Compressor(dest io.Writor, mediatype string) something like that. The compressor should be aligned with mediatype.

Reply:
@fuweid Should I update the option in this PR?

Reply:
Sure. No need to handle one thing in two prs🤣